service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/DiagnosticReportMapper.java

```diff
@@ -0,0 +1,184 @@
+
+        if (diagnosticReport.hasConclusion()) {
+            String narrativeStatementFromConclusion = buildNarrativeStatementForDiagnosticReport(
+                diagnosticReport, diagnosticReport.getConclusion()
+            );
+
+            reportLevelNarrativeStatements.append(narrativeStatementFromConclusion);
+        }
+
+        String codedDiagnosisText = diagnosticReport.getCodedDiagnosis().stream()
+            .map(this::extractTextFromCodedDiagnosisItems)
+            .flatMap(List::stream)
+            .collect(Collectors.joining(", "));
+
+        if (!codedDiagnosisText.isEmpty()) {
+            String narrativeStatementFromCodedDiagnosis = buildNarrativeStatementForDiagnosticReport(
+                diagnosticReport, codedDiagnosisText
+            );
+
+            reportLevelNarrativeStatements.append(narrativeStatementFromCodedDiagnosis);
+        }
+
+        diagnosticReportCompoundStatementTemplateParameters.narrativeStatements(reportLevelNarrativeStatements.toString());
+
+        return TemplateUtils.fillTemplate(
+            DIAGNOSTIC_REPORT_COMPOUND_STATEMENT_TEMPLATE,
+            diagnosticReportCompoundStatementTemplateParameters.build()
+        );
+    }
+
+    private List<Specimen> fetchSpecimens(DiagnosticReport diagnosticReport) {
+        if (!diagnosticReport.hasSpecimen()) {
+            return Collections.singletonList(generateDefaultSpecimen(diagnosticReport));
+        }
+
+        return diagnosticReport.getSpecimen()
+            .stream()
+            .map(specimenReference -> messageContext.getInputBundleHolder().getResource(specimenReference.getReferenceElement()))
+            .flatMap(Optional::stream)
+            .map(Specimen.class::cast)
+            .collect(Collectors.toList());
+    }
+
+    private Specimen generateDefaultSpecimen(DiagnosticReport diagnosticReport) {
+        Specimen specimen = new Specimen();
+
+        specimen.setId("Specimen/Default-1");
+
+        return specimen
+            .setAccessionIdentifier(new Identifier().setValue("DUMMY"))
+            .setCollection(new Specimen.SpecimenCollectionComponent().setCollected(new DateTimeType(diagnosticReport.getIssued())))
+            .setType(new CodeableConcept().setText("UNKNOWN"));
+    }
+
+    private List<Observation> fetchObservations(DiagnosticReport diagnosticReport) {
+        if (!diagnosticReport.hasResult()) {
+            return Collections.singletonList(generateDefaultObservation(diagnosticReport));
+        }
+
+        var inputBundleHolder = messageContext.getInputBundleHolder();
+        return diagnosticReport.getResult().stream()
+            .map(Reference::getReferenceElement)
+            .map(inputBundleHolder::getResource)
+            .flatMap(Optional::stream)
+            .map(Observation.class::cast)
+            .collect(Collectors.toList());
+    }
+
+    private Observation generateDefaultObservation(DiagnosticReport diagnosticReport) {
```

Comment:
I think it would be good to create static classes on for DefaultObservation with method get/build but as well with the methods to verify if the resource is a default resource. That could be used in ObservationMapper.prepareCommentType where we are verifying if the observation is empty report t should that this is something created by ourself before, not something that came with input bundle. Similar thing could be done with Specimen an well.

Reply:
I think that's a bit overkill for the purpose of this.
I can't see how this would be useful for `prepareCommentType` as well, I feel like it's just overcomplicating the logic where specimens or observations don't exist